client/asset/btc/btc.go

```diff
@@ -2361,6 +2419,429 @@ func (btc *baseWallet) fund(keep uint64, minConfs uint32, lockUnspents bool,
+			remainingOrders = newRemainingOrders
+			remainingIndexes = newRemainingIndexes
+			remainingUTXOs = utxoSetDiff(remainingUTXOs, fundingUTXOs)
+		}
+		allFundingUTXOs := make([][]*compositeUTXO, len(values))
+		for i := range values {
+			allFundingUTXOs[i] = indexToFundingCoins[i]
+		}
+		return allFundingUTXOs
+	}
+
+	fundInOrder := func(orderedValues []*asset.MultiOrderValue) [][]*compositeUTXO {
+		allFundingUTXOs := make([][]*compositeUTXO, 0, len(orderedValues))
+		remainingUTXOs := utxos
+		var totalFunded uint64
+		for _, value := range orderedValues {
+			enough := orderEnough(value.Value, value.MaxSwapCount, maxFeeRate, btc.initTxSizeBase, btc.initTxSize, btc.segwit, false)
+
+			var fundingUTXOs []*compositeUTXO
+			if maxLock > 0 {
+				if maxLock < totalFunded {
+					// Should never happen unless there is a bug in leastOverFundWithLimit
+					btc.log.Errorf("maxLock < totalFunded. %d < %d", maxLock, totalFunded)
+					return allFundingUTXOs
+				}
+				fundingUTXOs = leastOverFundWithLimit(enough, maxLock-totalFunded, remainingUTXOs)
+			} else {
+				fundingUTXOs = leastOverFund(enough, remainingUTXOs)
+			}
+			if len(fundingUTXOs) == 0 {
+				return allFundingUTXOs
+			}
+			totalFunded += sumUTXOs(fundingUTXOs)
+			if totalFunded > avail-keep {
+				return allFundingUTXOs
+			}
+			allFundingUTXOs = append(allFundingUTXOs, fundingUTXOs)
+			remainingUTXOs = utxoSetDiff(remainingUTXOs, fundingUTXOs)
+		}
+		return allFundingUTXOs
+	}
+
+	returnValues := func(allFundingUTXOs [][]*compositeUTXO) (coins []asset.Coins, redeemScripts [][]dex.Bytes, fundingCoins map[outPoint]*utxo, spents []*output, err error) {
+		coins = make([]asset.Coins, len(allFundingUTXOs))
+		fundingCoins = make(map[outPoint]*utxo)
+		spents = make([]*output, 0, len(allFundingUTXOs))
+		redeemScripts = make([][]dex.Bytes, len(allFundingUTXOs))
+		for i, fundingUTXOs := range allFundingUTXOs {
+			coins[i] = make(asset.Coins, len(fundingUTXOs))
+			redeemScripts[i] = make([]dex.Bytes, len(fundingUTXOs))
+			for j, output := range fundingUTXOs {
+				coins[i][j] = newOutput(output.txHash, output.vout, output.amount)
+				fundingCoins[outPoint{txHash: *output.txHash, vout: output.vout}] = &utxo{
+					txHash:  output.txHash,
+					vout:    output.vout,
+					amount:  output.amount,
+					address: output.address,
+				}
+				spents = append(spents, newOutput(output.txHash, output.vout, output.amount))
+				redeemScripts[i][j] = output.redeemScript
+			}
+		}
+		return
+	}
+
+	// Attempt to fund all orders by selecting the order that requires the least
+	// over funding, removing the funding utxos from the set of available utxos,
+	// and continuing until all orders are funded.
+	allFundingUTXOs := fundAllOrders()
+	if allFundingUTXOs != nil {
+		return returnValues(allFundingUTXOs)
+	}
+
+	// If could not fully fund, fund as much as possible in the priority
+	// order.
+	allFundingUTXOs = fundInOrder(values)
+	return returnValues(allFundingUTXOs)
```

Comment:
Good point @JoeGruffins.  @martonp has resolved this as suggested, so don't change anything now, but my only comment would have been regarding the semantics of the parameter being called `splitAllowed` instead of `allowPartialFallback` or similar.  As it stands, `fundMultiBestEffort` is not realistically going to be used from anything other than `fundMulti`, so I don't have a big issue with `splitAllowed`.